### PR TITLE
fix(root): entry-save validation and migration table-naming parity fixes

### PR DESCRIPTION
**Pull request by aqib-rx**
## Summary

First batch of the code-first ↔ Visual Schema Builder parity / data-loss fixes, opened early (and accumulating more commits on this same branch) so the review bots can surface feedback incrementally. **WIP — please review, but do not merge until the full plan lands on this branch.** Every fix is test-driven (failing test first, then the fix).

**1. `fix(admin)`: optional entry fields no longer block save when left blank**
- *Problem:* an optional field the user never filled blocked publish — empty optional email → "Invalid email address", empty optional password → min-length error, empty optional multi-select → "expected array, received string".
- *Root cause:* optional fields are built as `.optional().nullable()`, but the form seeds untouched string fields to `""`, and `.optional()` only tolerates `undefined`, so `""` still ran through `.email()` / `.min()` / `z.array()`.
- *Fix:* accept `""` alongside the field schema **without rewriting the value** (`z.union([z.literal(""), schema])`), reusing the idiom this file already used for the password field, and centralize it so the now-redundant password-specific union (and its unused `options` param) is removed.
- *Files:* `packages/admin/src/lib/field-validation.ts` (+ tests).

**2. `fix(admin)`: `hasMany` select renders as a real multi-value input**
- *Problem:* a multi-select could never be saved — it stored a scalar string and failed the field's `z.array()` schema (with a `<select multiple=false>` console warning).
- *Root cause:* `FieldRenderer` routed **every** `select` to the single-value `SelectInput`; no multi-value component existed. This is a shared-renderer bug, so it affected both code-first and builder-authored (`Allow Multiple`) fields.
- *Fix:* new `MultiSelectInput` that stores an array (seeds `[]`, never a scalar) with removable badges + an add dropdown, reusing the existing `@nextlyhq/ui` `Select*`/`Badge` primitives (no new deps); `FieldRenderer` dispatches to it when `field.hasMany` is set.
- *Files:* new `MultiSelectInput.tsx` (+ test), `FieldRenderer.tsx`, `SelectInput.tsx` (comment only).

**3. `fix(nextly)`: consistent table-name prefixing in `migrate:create` / `migrate:check`**
- *Problem:* `migrate:create` emitted `CREATE TABLE forms` for the form-builder plugin while the runtime creates `dc_forms`, so generated migrations diverged from the live database.
- *Root cause:* `toMinimalEntities` computed `tableName: e.dbName ?? ` + "`${prefix}${slug}`" + `, skipping the `dc_`/`single_`/`comp_` prefix whenever a plugin set a bare `dbName` (e.g. `"forms"`).
- *Fix:* a shared `resolvePrefixedTableName(slug, dbName, prefix)` generalizes the existing `resolveCollectionTableName` rule; both mirrored CLI sites route through it, and `resolveCollectionTableName` now delegates to it (behavior identical).
- *Files:* `resolve-table-name.ts` (+ test), `cli/commands/migrate-create.ts`, `cli/commands/migrate-check.ts`.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would change existing behavior)
- [ ] Documentation update
- [ ] Refactor / chore (no user-facing change)

## Related issues

Part of the code-first ↔ Visual Schema Builder parity effort. No tracking issue.

## Changeset

- [x] I added a changeset (one `patch` changeset covering all published packages, alpha lockstep)
- [x] I selected the correct semver bump (patch / minor / major)

## Test plan

- [x] `pnpm lint` (`@nextlyhq/admin` and `nextly`, `--max-warnings 0`)
- [x] `pnpm check-types` (both packages)
- [x] `pnpm build` (ran via the pre-push hook — 15/15 tasks)
- [ ] Manually verified in-browser (new behavior is covered by unit tests; a live admin re-verification is pending a dist rebuild)
- [x] New unit tests added and passing for each fix

> **Pre-existing baseline:** the `@nextlyhq/admin` suite has an unrelated, pre-existing set of failing tests (dashboard / role-management / schema-builder-settings / search-bar suites). None of those files import the modules changed here, and the failure set is identical with and without this PR.

## Checklist

- [x] My commits follow the [Conventional Commits](https://www.conventionalcommits.org/) spec (enforced by commitlint)
- [x] I targeted the `main` branch
- [ ] I updated relevant documentation (no doc changes needed for these fixes)

## Notes for reviewers

- **WIP:** this branch will accumulate the rest of the plan — component number column typing, builder-apply scoping + replayable migrations, the false SQLite type-change warning, code-first snapshot cross-contamination, and product-gated parity options. Please review incrementally; do not merge until it is complete.
- Design decisions that may pre-empt questions:
  - Empty-string tolerance is centralized in the optional branch of `generateClientSchema` rather than duplicated per type; the password-specific union was removed as redundant and its now-unused `options` param dropped.
  - `MultiSelectInput` intentionally reuses the same `Select*` / `Badge` primitives as `SelectInput` / `ChipsInput` for consistent light/dark styling and to avoid new dependencies.
  - `resolveCollectionTableName` now delegates to the new generic `resolvePrefixedTableName`; behavior is unchanged (covered by the new test) — the generalization exists only so the migration CLI can prefix `single_` / `comp_` entities too.

## Summary by CodeRabbit

* **New Features**
  * Added multi-select support for “has many” select fields, including removable badges and an add flow that hides already-selected options.
  * Schema apply now surfaces the exact DDL executed, and builder-run DDL can be saved for replay.
* **Bug Fixes**
  * Validation now tolerates empty-string for optional scalar fields while keeping array-backed selects strict; password empty-string behavior is more consistent.
  * Migration/code-first schema tooling now resolves physical table names using shared resolver rules.
  * Improved decimal field column typing and SQLite numeric typing/widening behavior.
* **Documentation**
  * Clarified single vs multi-select rendering behavior.
* **Tests**
  * Expanded coverage for multi-select interactions, validation edge cases, table-name resolvers, number/type classification, and locked-table operation filtering.

### CODE REVIEW

**Comment by mobeenabdullah** [comment]
@codex please review this PR

File: packages/nextly/src/domains/schema/utils/resolve-table-name.ts [hunk 1]
```diff
@@ -14,7 +14,22 @@ export function resolveCollectionTableName(
...
+// an un-prefixed table that diverges from the runtime's `dc_forms`.
+export function resolvePrefixedTableName(
+  slug: string,
+  dbName: string | undefined,
+  prefix: "dc_" | "single_" | "comp_"
+): string {
+  if (dbName && dbName.startsWith(prefix)) return dbName;
   const base = dbName ?? slug;
-  return `dc_${base.replace(/-/g, "_")}`;
+  return `${prefix}${base.replace(/-/g, "_")}`;
```
Comment (mobeenabdullah) [inline packages/nextly/src/domains/schema/utils/resolve-table-name.ts]: **The prefix rule here doesn't match what the runtime actually does for `single_` and `comp_`.**

This is the one I'd most like resolved before this lands, because it's the core goal of the fix (CLI-generated names matching the live database), and right now it only fully achieves that for collections.

**Context — the runtime uses a stronger normalizer.** For singles, `resolveSingleTableName` (`packages/nextly/src/domains/singles/services/resolve-single-table-name.ts`) normalizes with:

```ts
input.toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_+|_+$/g, "")
```

Components build their name the same way, in two places — `init/reload-config.ts:454` and `di/register.ts:1617`.

This helper only does `.replace(/-/g, "_")`: no lowercasing, no collapsing of other non-alphanumeric characters, no trimming of stray underscores.

**What that means in practice** (all realistic inputs):

| entity | slug / dbName | runtime creates | this helper returns |
| --- | --- | --- | --- |
| single | slug `siteSettings` | `single_sitesettings` | `single_siteSettings` |
| single | dbName `Site Settings` | `single_site_settings` | `single_Site Settings` |
| single | dbName `single_Foo` | `single_foo` | `single_Foo` (returned verbatim by the early return on line 32) |
| component | slug `Hero.Banner` | `comp_hero_banner` | `comp_Hero.Banner` |

A camelCase single slug is a very ordinary thing to write, so this isn't an exotic edge case.

To be clear about what this PR *does* fix: for a bare `dbName` like `"forms"` it's a real improvement (`forms` → `dc_forms`), and the collection path is genuinely correct because `resolveCollectionTableName` delegates here and its behaviour is unchanged. The gap is specifically that generalizing to `single_`/`comp_` adopted the *collection* normalizer rather than the runtime's.

**Suggested fix.** Use the same normalizer the runtime uses, and normalize before the prefix check so a prefixed-but-unnormalized `dbName` converges too:

```ts
function normalizeIdentifier(input: string): string {
  return input
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "");
}

export function resolvePrefixedTableName(
  slug: string,
  dbName: string | undefined,
  prefix: "dc_" | "single_" | "comp_"
): string {
  const base = normalizeIdentifier(dbName ?? slug);
  return base.startsWith(prefix) ? base : `${prefix}${base}`;
}
```

**One thing to check before you do that:** this would also change `resolveCollectionTableName`, which currently preserves case (`dbName: "MyTable"` → `dc_MyTable`). That's a behaviour change for collections, so it's worth confirming against the runtime collection path and the existing tests rather than assuming it's safe — it may deserve its own commit with the reasoning written down.

**Bigger picture, entirely your call:** there are now four sources of truth for this (`resolveCollectionTableName`, `resolveSingleTableName`, this helper, plus the two inline `comp_` constructions). Since you're already unifying, having `resolveSingleTableName` and the two `comp_` sites delegate to this one helper would mean the rule genuinely exists once. That may be more than you want in this PR — a follow-up is completely reasonable — but it would make the parity guarantee real rather than convention.

File: packages/admin/src/components/features/entries/fields/selection/MultiSelectInput.tsx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+              variant="default"
+              className="flex items-center gap-1 pr-1"
+            >
+              <span>{labelFor(optionValue)}</span>
+              {isInteractive && (
+                <button
+                  type="button"
+                  onClick={() => removeValue(optionValue)}
+                  className="ml-0.5 rounded-none p-0.5 hover:bg-black/10 focus:outline-none"
+                  tabIndex={-1}
```
Comment (mobeenabdullah) [inline packages/admin/src/components/features/entries/fields/selection/MultiSelectInput.tsx]: **Keyboard users can't remove a selected value here.**

`tabIndex={-1}` takes this × button out of the tab order, so the only way to remove a chosen option is a mouse click.

**Why this happened is worth explaining, because you followed the right instinct.** This markup mirrors `ChipsInput.tsx:124-135`, which uses exactly the same classes and the same `tabIndex={-1}`. Copying the sibling component was reasonable. The difference is that `ChipsInput` also has a **text input** with a keyboard handler:

```ts
// ChipsInput.tsx:99
} else if (e.key === "Backspace" && inputValue === "" && chips.length > 0) {
```

So in `ChipsInput` a keyboard user still has a removal path — Backspace on the empty input — and `tabIndex={-1}` merely stops the × cluttering the tab order.

`MultiSelectInput` has no text input and no key handler. Its only interactive elements are the Select trigger and these × buttons, so with `tabIndex={-1}` a keyboard-only user can add values but can **never remove one**. The escape hatch that made the pattern safe in `ChipsInput` doesn't exist here.

**Suggested fix — drop `tabIndex={-1}` and give the button a visible focus ring** (`focus:outline-none` on its own removes the focus indicator, which compounds the problem):

```tsx
<button
  type="button"
  onClick={() => removeValue(optionValue)}
  className="ml-0.5 rounded-none p-0.5 hover:bg-[var(--nx-surface-hover)] focus-visible:ring-1 focus-visible:ring-ring focus-visible:outline-none"
  aria-label={`Remove ${labelFor(optionValue)}`}
>
```

`ChipsInput` already uses `ring-ring` at line 108, so that token is the established one here.

**Separately, on the same line: `hover:bg-black/10` is a hardcoded colour.** The admin styling rule in `AGENTS.md` is token-driven `--nx-*` with zero hardcoded colours and both modes working. In dark mode a 10% black wash over an already-dark badge is close to invisible, so the hover affordance effectively disappears.

This one is inherited from `ChipsInput`, so I don't think it's fair to call it a defect *you* introduced — but it does get copied into a second component here. Either swap it for a token now, or (equally fine) leave both and open a small follow-up that fixes `ChipsInput` and `MultiSelectInput` together. I'd just rather it be a decision than an accident.

File: packages/admin/src/components/features/entries/fields/selection/MultiSelectInput.tsx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    control,
+    // hasMany selects store an array. Seed [] (never a scalar) so the field's
+    // z.array() schema is satisfied for an untouched field and every update
+    // keeps the value an array.
+    defaultValue: (Array.isArray(field.defaultValue)
+      ? field.defaultValue
+      : []) as TFieldValues[Path<TFieldValues>],
+  });
+
+  const selected: string[] = Array.isArray(value) ? value : [];
```
Comment (mobeenabdullah) [inline packages/admin/src/components/features/entries/fields/selection/MultiSelectInput.tsx]: **A stored non-array value is silently discarded here, which could overwrite real data.**

`Array.isArray(value) ? value : []` means any stored value that isn't an array renders as *no selection*. If the user then saves the entry — even without touching this field — the scalar is replaced by `[]` and the original value is gone.

**Is that reachable?** I think it's narrow but real. It can't come from this form (the entries `getDefaultValues` seeds `hasMany` selects to `[]`, and the pre-fix bug blocked saving), but a scalar can land in the column from:

- rows written through the Direct API or REST before this field became `hasMany`
- a field switched from single-select to **Allow Multiple** in the Schema Builder, where existing rows still hold the old scalar
- imported or seeded data

That middle one feels most likely, and it's squarely inside the code-first ↔ builder parity work this PR belongs to.

**Suggested fix** — treat a non-empty scalar as a one-item selection instead of dropping it:

```ts
// A value stored before the field became hasMany (or written by the API) can be
// a scalar. Treat it as a single selection rather than discarding it, so saving
// an untouched field cannot erase the existing value.
const selected: string[] = Array.isArray(value)
  ? value
  : typeof value === "string" && value !== ""
    ? [value]
    : [];
```

If you'd rather not widen the behaviour, that's defensible too — but then it's worth a comment saying the scalar is dropped deliberately, so the next reader doesn't assume it's an oversight.

File: .changeset/entry-validation-and-table-naming-fixes.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+---
+"@nextlyhq/adapter-drizzle": patch
+"@nextlyhq/adapter-mysql": patch
+"@nextlyhq/adapter-postgres": patch
+"@nextlyhq/adapter-sqlite": patch
+"@nextlyhq/admin": patch
```
Comment (mobeenabdullah) [inline .changeset/entry-validation-and-table-naming-fixes.md:6]: **This changeset is missing `@nextlyhq/admin-css`.**

The list here has 18 packages; the canonical set is currently 19. Comparing against the most recent changeset on `main` (`.changeset/webhook-delivery-engine.md`), the missing entry is:

```
"@nextlyhq/admin-css": patch
```

`admin-css` is a real published package (`packages/admin-css/package.json`, currently `0.0.2-alpha.36`). It's easy to miss because it's new — it only joined the fixed group in #249.

**Why it matters:** all packages version in lockstep, so an omitted package doesn't get bumped with the others and drifts a version behind. This is the same class of problem #246 fixed from the other direction (a package listed that no longer existed).

**Fix:** add the line above, keeping the list alphabetical — it sits between `"@nextlyhq/adapter-sqlite"` and `"@nextlyhq/admin"`.

**Review by mobeenabdullah — COMMENTED** [review]
Reviewed this carefully end to end — read the full diff, then checked each claim against the surrounding code on your branch rather than taking the diff at face value. Overall this is high-quality work: three genuinely distinct bugs, each with a failing-test-first commit, real root-cause analysis in the description, and comments that explain *why* rather than *what*. The `resolveCollectionTableName` delegation preserving existing behaviour is a nice touch, and centralizing the empty-string tolerance instead of duplicating it per field type is the right call.

Six comments below. One I'd like resolved before merge, the rest are smaller or are questions.

**Things I checked that turned out fine** (recording these so nobody re-litigates them):

- I initially suspected the `z.union([z.literal(""), schema])` change would let `""` through for **number** fields and break numeric columns. It doesn't — `getDefaultValues` seeds numbers to `null`, not `""` (`useEntryForm.ts:311`), and `NumberInput` explicitly converts `""` back to `null` on change (`NumberInput.tsx:169`). So the union's reach is genuinely limited to string-ish fields. Good.
- Server-side validation tolerates `""` for optional fields (`entry-validation.ts:129`), so the fix can't cause a save to fail on the server.
- `resolveCollectionTableName` really is behaviour-identical after the refactor.
- The `hasMany` renderer fix *does* reach Singles, because `SingleForm` renders through `EntryFormContent` rather than its own field renderer.

**One gap that isn't visible in this diff**, so I couldn't leave it inline:

`SingleForm.tsx:198-206` has its own `getDefaultValues` that lumps `select` and `radio` in with the text types and seeds them to `""`:

```ts
case "select":
case "radio":
  defaults[fieldName] = (field as { defaultValue?: string }).defaultValue ?? "";
```

The entries version (`useEntryForm.ts:320-334`) handles `hasMany` properly and seeds `[]`. Because Singles now render through your new `MultiSelectInput` but still seed `""`, an untouched `hasMany` select on a Single submits a scalar `""` where an array is expected. With this PR's union change that now passes client validation instead of being rejected, so it reaches the server rather than erroring in the browser.

Since this branch is explicitly the parity effort, aligning `SingleForm`'s seeding with the entries version looks like it belongs in scope — but you know the full plan better than I do, so treat it as a suggestion for this branch or a follow-up.

Not approving or requesting changes formally, since you've marked this WIP and asked for incremental review. Happy to re-review any of these once you've had a look — and push back on anything you disagree with, particularly the table-name one, where you may have context on `dbName` semantics that I don't.

File: packages/admin/src/lib/field-validation.ts [hunk 5]
```diff
@@ -344,7 +344,16 @@ function fieldToZodSchema(
...
   if (!isRequired) {
-    schema = schema.optional().nullable();
+    // An untouched optional field arrives from the form as "" (empty string).
+    // `.optional()` only tolerates `undefined`, so "" would otherwise be run
+    // through the field's format/length/array rules (email, minLength, array)
+    // and wrongly block save. Accept "" alongside the field schema — without
+    // rewriting the value — so an empty optional field is valid while a typed
+    // value is still validated.
+    schema = z
+      .union([z.literal(""), schema])
```
Comment (mobeenabdullah) [inline packages/admin/src/lib/field-validation.ts]: **Question, not a blocker — is `""` the value you want to reach the database?**

The union deliberately doesn't rewrite the value (your comment says so explicitly, and the reasoning is sound), so an untouched optional text/email field now validates *and submits* as `""` rather than being omitted or nulled.

I checked the server side and validation is genuinely fine with this: `isEmpty()` in `packages/nextly/src/shared/lib/entry-validation.ts:129-135` counts `""` as empty, and empty values only ever trip required-ness — format and length rules are skipped. So nothing rejects the save. This is not a correctness bug.

The part I'd like your read on is what ends up **stored**. A nullable column will now hold `""` for a field the user never filled, so the same "empty" state can be represented two ways depending on how the row was written (`""` from the admin, `NULL` from the API or a migration). That makes `WHERE col IS NULL` quietly incomplete, and it's the kind of inconsistency that's much cheaper to settle now than after real data exists.

Two reasonable directions:

1. **Keep as-is** — simplest, and honest to "the user submitted an empty field". Worth a sentence in the comment noting `""` is what gets persisted, so it's a recorded decision.
2. **Normalize on the way out** — strip empty strings for optional fields when building the submit payload (not in the schema, so validation stays as you designed it), letting the column stay `NULL`.

I lean slightly toward 2 for data hygiene, but you're closer to this area than I am and either is defensible. Mostly I want to make sure it's a choice rather than a side effect.

**Comment by mobeenabdullah** [comment]
@codex please review this PR.

**Comment by mobeenabdullah** [comment]
@codex please review this PR.

**Review by faisal-rx — APPROVED** [review]
